From 0d15bb2998608945383326d478cca4791e2bc4dc Mon Sep 17 00:00:00 2001
From: Motomu Utsumi <motomuman@google.com>
Date: Fri, 17 Feb 2023 11:52:14 +0900
Subject: [PATCH] cronet: Fix ListenerLast API linter error

Update UrlRequest.Builder#newUrlRequestBuilder and
UrlRequest.Builder#newBidirectionalStreamBuilder
Also, update UrlRequestBuilderImpl and newBidirectionalStreamBuilder
arguments order to be aligned.

Bug: 265674359
Test: TH

Change-Id: Ie5ad5072f356ebc6a72a5b9915fd30c1f8e887d9
---
 .../net/http/ExperimentalHttpEngine.java      |  4 +-
 .../api/src/android/net/http/HttpEngine.java  | 50 +++++++++++++++++--
 .../chromium/net/impl/CronetEngineBase.java   |  2 +-
 .../net/impl/CronetUrlRequestContext.java     |  2 +-
 .../CronetHttpURLConnection.java              |  2 +-
 5 files changed, 51 insertions(+), 9 deletions(-)

diff --git a/components/cronet/android/api/src/android/net/http/ExperimentalHttpEngine.java b/components/cronet/android/api/src/android/net/http/ExperimentalHttpEngine.java
index 8a195f21..4f3f74a1 100644
--- a/components/cronet/android/api/src/android/net/http/ExperimentalHttpEngine.java
+++ b/components/cronet/android/api/src/android/net/http/ExperimentalHttpEngine.java
@@ -502,11 +502,11 @@ public abstract class ExperimentalHttpEngine extends HttpEngine {
 
     @Override
     public abstract ExperimentalBidirectionalStream.Builder newBidirectionalStreamBuilder(
-            String url, BidirectionalStream.Callback callback, Executor executor);
+            String url, Executor executor, BidirectionalStream.Callback callback);
 
     @Override
     public abstract ExperimentalUrlRequest.Builder newUrlRequestBuilder(
-            String url, UrlRequest.Callback callback, Executor executor);
+            String url, Executor executor, UrlRequest.Callback callback);
 
     /**
      * Starts NetLog logging to a specified directory with a bounded size. The NetLog will contain
diff --git a/components/cronet/android/api/src/android/net/http/HttpEngine.java b/components/cronet/android/api/src/android/net/http/HttpEngine.java
index 27a0da94..8b6d93ae 100644
--- a/components/cronet/android/api/src/android/net/http/HttpEngine.java
+++ b/components/cronet/android/api/src/android/net/http/HttpEngine.java
@@ -524,11 +524,31 @@ public abstract class HttpEngine {
      * operations and causing exceptions during shutdown.
      *
      * @param url URL for the generated requests.
-     * @param callback callback object that gets invoked on different events.
      * @param executor {@link Executor} on which all callbacks will be invoked.
+     * @param callback callback object that gets invoked on different events.
      */
     public abstract UrlRequest.Builder newUrlRequestBuilder(
-            String url, UrlRequest.Callback callback, Executor executor);
+            String url, Executor executor, UrlRequest.Callback callback);
+
+    /**
+     * Creates a builder for {@link UrlRequest}. All callbacks for
+     * generated {@link UrlRequest} objects will be invoked on
+     * {@code executor}'s threads. {@code executor} must not run tasks on the
+     * thread calling {@link Executor#execute} to prevent blocking networking
+     * operations and causing exceptions during shutdown.
+     *
+     * @param url URL for the generated requests.
+     * @param callback callback object that gets invoked on different events.
+     * @param executor {@link Executor} on which all callbacks will be invoked.
+     */
+    // This API is kept for the backward compatibility in upstream
+    // TODO(motomuman) Hide this API
+    // This API is not hidden since this API is used in internal master and removing this makes
+    // presubmit fail. Once internal use is replaced by above API, this API will be hidden.
+    public UrlRequest.Builder newUrlRequestBuilder(String url, UrlRequest.Callback callback,
+            @SuppressLint("ListenerLast") Executor executor) {
+        return newUrlRequestBuilder(url, executor, callback);
+    }
 
     /**
      * Creates a builder for {@link BidirectionalStream} objects. All callbacks for
@@ -538,12 +558,34 @@ public abstract class HttpEngine {
      * may be thrown at shutdown time.
      *
      * @param url URL for the generated streams.
+     * @param executor the {@link Executor} on which {@code callback} methods will be invoked.
      * @param callback the {@link BidirectionalStream.Callback} object that gets invoked upon
      * different events occurring.
-     * @param executor the {@link Executor} on which {@code callback} methods will be invoked.
      *
      * @return the created builder.
      */
     public abstract BidirectionalStream.Builder newBidirectionalStreamBuilder(
-            String url, BidirectionalStream.Callback callback, Executor executor);
+            String url, Executor executor, BidirectionalStream.Callback callback);
+
+    /**
+     * Creates a builder for {@link BidirectionalStream} objects. All callbacks for
+     * generated {@code BidirectionalStream} objects will be invoked on
+     * {@code executor}. {@code executor} must not run tasks on the
+     * current thread, otherwise the networking operations may block and exceptions
+     * may be thrown at shutdown time.
+     *
+     * @param url URL for the generated streams.
+     * @param callback the {@link BidirectionalStream.Callback} object that gets invoked upon
+     * different events occurring.
+     * @param executor the {@link Executor} on which {@code callback} methods will be invoked.
+     *
+     * @return the created builder.
+     *
+     * @hide
+     */
+    // This API is kept for the backward compatibility in upstream
+    public BidirectionalStream.Builder newBidirectionalStreamBuilder(
+            String url, BidirectionalStream.Callback callback, Executor executor) {
+        return newBidirectionalStreamBuilder(url, executor, callback);
+    }
 }
diff --git a/components/cronet/android/java/src/org/chromium/net/impl/CronetEngineBase.java b/components/cronet/android/java/src/org/chromium/net/impl/CronetEngineBase.java
index 47d60752..2a636ee0 100644
--- a/components/cronet/android/java/src/org/chromium/net/impl/CronetEngineBase.java
+++ b/components/cronet/android/java/src/org/chromium/net/impl/CronetEngineBase.java
@@ -109,7 +109,7 @@ public abstract class CronetEngineBase extends ExperimentalHttpEngine {
 
     @Override
     public ExperimentalUrlRequest.Builder newUrlRequestBuilder(
-            String url, UrlRequest.Callback callback, Executor executor) {
+            String url, Executor executor, UrlRequest.Callback callback) {
         return new UrlRequestBuilderImpl(url, callback, executor, this);
     }
 
diff --git a/components/cronet/android/java/src/org/chromium/net/impl/CronetUrlRequestContext.java b/components/cronet/android/java/src/org/chromium/net/impl/CronetUrlRequestContext.java
index 4afac795..077f1271 100644
--- a/components/cronet/android/java/src/org/chromium/net/impl/CronetUrlRequestContext.java
+++ b/components/cronet/android/java/src/org/chromium/net/impl/CronetUrlRequestContext.java
@@ -274,7 +274,7 @@ public class CronetUrlRequestContext extends CronetEngineBase {
 
     @Override
     public ExperimentalBidirectionalStream.Builder newBidirectionalStreamBuilder(
-            String url, BidirectionalStream.Callback callback, Executor executor) {
+            String url, Executor executor, BidirectionalStream.Callback callback) {
         return new BidirectionalStreamBuilderImpl(url, callback, executor, this);
     }
 
diff --git a/components/cronet/android/java/src/org/chromium/net/urlconnection/CronetHttpURLConnection.java b/components/cronet/android/java/src/org/chromium/net/urlconnection/CronetHttpURLConnection.java
index 8fe5bc27..85bcce7f 100644
--- a/components/cronet/android/java/src/org/chromium/net/urlconnection/CronetHttpURLConnection.java
+++ b/components/cronet/android/java/src/org/chromium/net/urlconnection/CronetHttpURLConnection.java
@@ -257,7 +257,7 @@ public class CronetHttpURLConnection extends HttpURLConnection {
         }
         final ExperimentalUrlRequest.Builder requestBuilder =
                 (ExperimentalUrlRequest.Builder) mCronetEngine.newUrlRequestBuilder(
-                        getURL().toString(), new CronetUrlRequestCallback(), mMessageLoop);
+                        getURL().toString(), mMessageLoop, new CronetUrlRequestCallback());
         if (doOutput) {
             if (method.equals("GET")) {
                 method = "POST";
-- 
2.39.2.637.g21b0678d19-goog

